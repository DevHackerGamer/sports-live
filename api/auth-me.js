const { getUserType } = require('../lib/auth');

// Unified handler: resolves user type via Clerk when available, else allows dev header override
module.exports = async function handler(req, res) {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Methods', 'GET, OPTIONS');
  res.setHeader('Access-Control-Allow-Headers', 'Content-Type, Authorization, X-User-Type');

  if (req.method === 'OPTIONS') return res.status(200).end();
  if (req.method !== 'GET') return res.status(405).json({ success: false, error: 'Method not allowed' });

  try {
    const type = (await getUserType(req)) || 'user';
    const isAdmin = String(type).toLowerCase() === 'admin';
    return res.status(200).json({ success: true, type, isAdmin });
  } catch (error) {
    console.error('auth-me error:', error);
    return res.status(200).json({ success: true, type: 'user', isAdmin: false });
  }
};
