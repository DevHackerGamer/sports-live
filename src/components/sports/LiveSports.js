import React from 'react';
import { useLiveSports } from '../../hooks/useLiveSports';
import '../../styles/LiveSports.css';



const LiveSports = () => {
  // Update time every 10ms
  const [currentTime, setCurrentTime] = React.useState(new Date());
  React.useEffect(() => {
    const timer = setInterval(() => {
      setCurrentTime(new Date());
    }, 10);
    return () => clearInterval(timer);
  }, []);
  const { 
    sportsData, 
    isConnected, 
    error, 
    lastUpdated, 
    refreshData 
  } = useLiveSports();

  const formatTime = (date) => {
    if (!date) return '';
    return date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
  };

  const formatDate = (dateString) => {
    if (!dateString) return '';
    const date = new Date(dateString);
    return date.toLocaleDateString([], { 
      weekday: 'short', 
      month: 'short', 
      day: 'numeric' 
    });
  };

  const formatShortDate = (dateString) => {
    if (!dateString) return '';
    const date = new Date(dateString);
    return date.toLocaleDateString([], { month: 'short', day: 'numeric' });
  };

  const getStatusBadge = (status) => {
    const statusConfig = {
      live: { text: 'LIVE', className: 'status-live' },
      final: { text: 'FINAL', className: 'status-final' },
      scheduled: { text: 'SCHEDULED', className: 'status-scheduled' },
      default: { text: 'UPCOMING', className: 'status-default' }
    };
    
    const config = statusConfig[status] || statusConfig.default;
    return <span className={`status-badge ${config.className}`}>{config.text}</span>;
  };

  if (error) {
    return (
      <div className="live-sports error-state">
        <div className="error-content">
          <h3>Connection Error</h3>
          <p>Unable to fetch live sports data</p>
          <button onClick={refreshData} className="retry-button">
            Retry
          </button>
        </div>
      </div>
    );
  }

  if (!sportsData) {
    return (
      <div className="live-sports loading-state">
        <div className="loading-content">
          <div className="loading-spinner"></div>
          <h3>Loading live matches...</h3>
        </div>
      </div>
    );
  }

  return (
    <div className="live-sports">
      <div className="sports-header">
        <div className="header-left">
          <h2>Live Football</h2>
          <div className="connection-indicator">
            <div className={`status-dot ${isConnected ? 'connected' : 'disconnected'}`}></div>
            <span className="status-text">
              {isConnected ? 'Live' : 'Offline'}
            </span>
          </div>
        </div>

        <div className="header-right">
          <div className="current-time">
            {currentTime.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit', second: '2-digit' })}
          </div>
          {lastUpdated && (
            <div className="last-updated">
              Updated {formatTime(lastUpdated)}
            </div>
          )}
        </div>
      </div>

      {/* Optional context about the fetched window across competitions */}
      {(sportsData.dateFrom || sportsData.dateTo || sportsData.totalMatches) && (
        <div className="fetch-context" style={{ display: 'flex', gap: 12, alignItems: 'center', margin: '8px 0' }}>
          {sportsData.dateFrom && (
            <span className="date-range">Range: {formatShortDate(sportsData.dateFrom)} â†’ {formatShortDate(sportsData.dateTo)}</span>
          )}
          {typeof sportsData.totalMatches === 'number' && (
            <span className="count">Matches: {sportsData.totalMatches}</span>
          )}
          {sportsData.environment && (
            <span className="env">Env: {sportsData.environment}</span>
          )}
        </div>
      )}

      <div className="matches-grid">
        {sportsData.games && sportsData.games.length > 0 ? (
          sportsData.games.map((game) => (
            <div key={game.id} className="match-card">
              <div className="match-header">
                <span className="competition">
                  {game.competition}
                  {game.competitionCode && (
                    <span className="competition-code" style={{ marginLeft: 8, fontSize: 12, opacity: 0.8 }}>
                      [{game.competitionCode}]
                    </span>
                  )}
                </span>
                {getStatusBadge(game.status)}
              </div>
              
              <div className="match-teams">
                <div className="team">
                  <span className="team-name">{game.homeTeam}</span>
                  <span className="team-score">{game.homeScore}</span>
                </div>
                
                <div className="match-separator">vs</div>
                
                <div className="team">
                  <span className="team-name">{game.awayTeam}</span>
                  <span className="team-score">{game.awayScore}</span>
                </div>
              </div>

              <div className="match-details">
                {game.minute && game.status === 'live' && (
                  <span className="match-time">{game.minute}'</span>
                )}
                {game.matchday && (
                  <span className="matchday" style={{ marginLeft: 8 }}>MD {game.matchday}</span>
                )}
                {game.venue && game.venue !== 'TBD' && (
                  <span className="venue">{game.venue}</span>
                )}
                {game.utcDate && game.status === 'scheduled' && (
                  <span className="scheduled-time">
                    {formatDate(game.utcDate)} at {formatTime(new Date(game.utcDate))}
                  </span>
                )}
              </div>
            </div>
          ))
        ) : (
          <div className="no-matches">
            <p>No matches available</p>
            <button onClick={refreshData} className="refresh-btn">
              Refresh
            </button>
          </div>
        )}
      </div>

      <div className="sports-footer">
        <button onClick={refreshData} className="refresh-btn">
          Refresh Data
        </button>
        <div className="data-info">
          <span className="source">Source: {sportsData?.source || 'Unknown'}</span>
          {sportsData?.totalMatches && (
            <span className="count">{sportsData.totalMatches} matches</span>
          )}
        </div>
      </div>
    </div>
  );
};

export default LiveSports;
