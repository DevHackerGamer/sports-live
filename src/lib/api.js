// API client for making HTTP requests to our MongoDB-based backend
// This replaces Firebase real-time database functionality

class ApiClient {
  constructor(baseUrl = '') {
    this.baseUrl = baseUrl;
  }

  async request(endpoint, options = {}) {
    const url = `${this.baseUrl}${endpoint}`;
    const config = {
      headers: {
        'Content-Type': 'application/json',
  // Optional dev pass-through; in production backend should validate Clerk token
  ...(options.userType ? { 'X-User-Type': options.userType } : {}),
        ...options.headers,
      },
      ...options,
    };

    if (config.body && typeof config.body === 'object') {
      config.body = JSON.stringify(config.body);
    }

    try {
      const response = await fetch(url, config);
      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.message || `HTTP ${response.status}`);
      }

      return data;
    } catch (error) {
      console.error(`API request failed: ${url}`, error);
      throw error;
    }
  }

  // Matches API
  async getMatches() {
  // Request a broader upcoming window explicitly
  return this.request('/api/matches?range=10&includePast=&limit=500');
  }

  async getMatchesByDate(dateFrom, dateTo, limit = 1000) {
    const params = new URLSearchParams({ dateFrom, dateTo, limit: String(limit) });
    return this.request(`/api/matches?${params.toString()}`);
  }

  async getMatch(id) {
    return this.request(`/api/matches/${id}`);
  }
    // Alias for compatibility
    async getMatchById(id) {
      return this.getMatch(id);
    }
  
  async getMatchEvents(id) {
    return this.request(`/api/matches/${id}/events`);
  }
  
  async addMatchEvent(id, event, options = {}) {
    return this.request(`/api/matches/${id}/events`, {
      method: 'POST',
      body: event,
      ...options,
    });
  }
  
  async updateMatchEvent(id, eventId, updates, options = {}) {
    return this.request(`/api/matches/${id}/events/${eventId}`, {
      method: 'PUT',
      body: updates,
      ...options,
    });
  }
  
  async deleteMatchEvent(id, eventId, options = {}) {
    return this.request(`/api/matches/${id}/events/${eventId}`, {
      method: 'DELETE',
      ...options,
    });
  }
  
  // Update a match (accepts optional fetch options like headers)
  async updateMatch(id, updates, options = {}) {
    return this.request(`/api/matches/${id}`, {
      method: 'PUT',
      body: updates,
      ...options,
    });
  }

  // Get all news or by league
  async getFootballNews(leagueCode = null, limit = 20) {
    let url = `/api/football-news?limit=${limit}`;
    if (leagueCode) url += `&leagueCode=${leagueCode}`;
    return this.request(url);
  }
// Get football highlights
async getFootballHighlights(leagueName) {
  return this.request(`/api/football-highlights?leagueName=${encodeURIComponent(leagueName)}`);
}


  async createMatches(matches) {
    return this.request('/api/matches', {
      method: 'POST',
      body: { matches },
    });
  }

  // (Legacy duplicate removed â€“ single updateMatch above handles this)

  async deleteMatch(id) {
    return this.request(`/api/matches/${id}`, {
      method: 'DELETE',
    });
  }

  // Match Statistics API
  async getMatchStatistics(matchId) {
    return this.request(`/api/match-statistics?matchId=${encodeURIComponent(matchId)}`);
  }

  async createMatchStatistics(matchId, statistics) {
    return this.request('/api/match-statistics', {
      method: 'POST',
      body: { matchId, ...statistics },
    });
  }

  async updateMatchStatistics(matchId, statistics) {
    return this.request('/api/match-statistics', {
      method: 'PUT',
      body: { matchId, ...statistics },
    });
  }

  async deleteMatchStatistics(matchId) {
    return this.request(`/api/match-statistics?matchId=${encodeURIComponent(matchId)}`, {
      method: 'DELETE',
    });
  }

  // Reports API - recognize change`
async getReports(id) {
  if (id) {
    return this.request(`/api/reporting/${id}`);
  }
  return this.request('/api/reporting');
}

async createReport(report) {
  return this.request('/api/reporting', {
    method: 'POST',
    body: report,
  });
}

async updateReport(id, updates) {
  return this.request(`/api/reporting/${id}`, {
    method: 'PUT',
    body: updates,
  });
}

async deleteReport(id) {
  return this.request(`/api/reporting/${id}`, {
    method: 'DELETE',
  });
}


  // Teams API
  async getTeams() {
    return this.request('/api/teams');
  }

  async createTeams(teams) {
    return this.request('/api/teams', {
      method: 'POST',
      body: { teams },
    });
  }
  async getTeamById(id) {
  return this.request(`/api/teams?id=${id}`);
}

  //standings API

  // Standings API
    async getStandings({ competition, season, type, stage, limit = 20 } = {}) {
    const params = new URLSearchParams();
    if (competition) params.append('competition', competition);
    if (season) params.append('season', season);
    if (type) params.append('type', type);
    if (stage) params.append('stage', stage);
    params.append('limit', limit);

    return this.request(`/api/standings?${params.toString()}`);
  }

  async getStandingById(id) {
    return this.request(`/api/standings/${id}`);
  }

  // Users API
  async getUserFavorites(userId) {
    return this.request(`/api/users/${userId}/favorites`);
  }

  async addUserFavorite(userId, teamName) {
    return this.request(`/api/users/${userId}/favorites`, {
      method: 'POST',
      body: { teamName },
    });
  }

  async removeUserFavorite(userId, teamName) {
  const encoded = encodeURIComponent(teamName);
  return this.request(`/api/users/${userId}/favorites/${encoded}`, {
      method: 'DELETE',
    });
  }

  async updateUserFavorites(userId, favorites) {
    return this.request(`/api/users/${userId}/favorites`, {
      method: 'PUT',
      body: { favorites },
    });
  }

  async getEventLog(params = {}) {
    const qs = new URLSearchParams();
    if (params.limit) qs.append('limit', String(params.limit));
    if (params.type) qs.append('type', params.type);
    if (params.matchId) qs.append('matchId', params.matchId);
    if (params.startDate) qs.append('startDate', params.startDate);
    if (params.endDate) qs.append('endDate', params.endDate);
    const query = qs.toString();
    return this.request(`/api/event-log${query ? '?' + query : ''}`);
  }

  // User Watchlist (Matches)
  async getUserWatchlist(userId) {
    const params = new URLSearchParams();
    if (userId) params.append('userId', userId);
    return this.request(`/api/user-matches?${params.toString()}`);
  }

  async addUserMatch(userId, match) {
    const matchId = match?.id || match?._id || match?.matchId;
    if (!matchId) throw new Error('matchId required');
    return this.request('/api/user-matches', {
      method: 'POST',
      body: { userId, matchId, match },
    });
  }

  async removeUserMatch(userId, matchId) {
    const qs = new URLSearchParams({ userId, matchId: String(matchId) });
    return this.request(`/api/user-matches?${qs.toString()}`, { method: 'DELETE' });
  }
}

// Create a singleton instance
const apiClient = new ApiClient();

// Real-time data simulation using polling
class RealTimeData {
  constructor(apiClient) {
    this.apiClient = apiClient;
    this.subscriptions = new Map();
    this.intervals = new Map();
  }

  // Simulate Firebase onValue with polling
  onValue(endpoint, callback, errorCallback, pollInterval = 30000) {
    const id = Math.random().toString(36);
    
    // Initial fetch
    this.fetchAndNotify(endpoint, callback, errorCallback);
    
    // Set up polling
    const intervalId = setInterval(() => {
      this.fetchAndNotify(endpoint, callback, errorCallback);
    }, pollInterval);
    
    this.intervals.set(id, intervalId);
    this.subscriptions.set(id, { endpoint, callback, errorCallback });
    
    // Return unsubscribe function
    return () => {
      if (this.intervals.has(id)) {
        clearInterval(this.intervals.get(id));
        this.intervals.delete(id);
      }
      this.subscriptions.delete(id);
    };
  }

  async fetchAndNotify(endpoint, callback, errorCallback) {
    try {
      let data;
      switch (endpoint) {
        case 'matches':
          data = await this.apiClient.getMatches();
          // Convert array to object with IDs as keys (Firebase-like structure)
          const matchesObj = {};
          data.data.forEach(match => {
            matchesObj[match.id] = match;
          });
          callback({ val: () => matchesObj });
          break;
        case 'teams':
          data = await this.apiClient.getTeams();
          callback({ val: () => data.data });
          break;
        default:
          if (endpoint.startsWith('users/') && endpoint.includes('/favorites')) {
            const userId = endpoint.split('/')[1];
            data = await this.apiClient.getUserFavorites(userId);
            callback({ val: () => data.data });
          }
          break;
      }
    } catch (error) {
      console.error('Error fetching data:', error);
      if (errorCallback) {
        errorCallback(error);
      }
    }
  }

  // Clean up all subscriptions
  cleanup() {
    this.intervals.forEach(intervalId => clearInterval(intervalId));
    this.intervals.clear();
    this.subscriptions.clear();
  }
}

const realTimeData = new RealTimeData(apiClient);

// Firebase-compatible API
export const db = {
  // Placeholder for compatibility
};

export const ref = (db, path) => {
  return { path };
};

export const onValue = (ref, callback, errorCallback) => {
  const path = ref.path;
  return realTimeData.onValue(path, callback, errorCallback);
};

export const update = async (ref, data) => {
  const path = ref.path;
  
  if (path.startsWith('users/') && path.includes('/favorites')) {
    const userId = path.split('/')[1];
    await apiClient.updateUserFavorites(userId, data.favorites);
  } else if (path === 'matches') {
    // Handle bulk match updates
    const matchesArray = Object.values(data);
    await apiClient.createMatches(matchesArray);
  } else if (path === 'teams') {
    await apiClient.createTeams(data);
  }
};

export const set = async (ref, data) => {
  return update(ref, data);
};

export const get = async (ref) => {
  const path = ref.path;
  
  try {
    let data;
    switch (path) {
      case 'matches':
        data = await apiClient.getMatches();
        return { val: () => data.data };
      case 'teams':
        data = await apiClient.getTeams();
        return { val: () => data.data };
      default:
        if (path.startsWith('users/') && path.includes('/favorites')) {
          const userId = path.split('/')[1];
          data = await apiClient.getUserFavorites(userId);
          return { val: () => data.data };
        }
        return { val: () => null };
    }
  } catch (error) {
    console.error('Error getting data:', error);
    return { val: () => null };
  }
};

export const child = (ref, path) => {
  return { path: `${ref.path}/${path}` };
};

// Export API client for direct use
export { apiClient, realTimeData };
// Export getMatchById for direct use
export const getMatchById = (id) => apiClient.getMatchById(id);

export default {
  db,
  ref,
  onValue,
  update,
  set,
  get,
  child,
  apiClient,
  realTimeData,
};
