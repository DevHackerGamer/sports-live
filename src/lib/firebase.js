// Firebase client initialization (App + Realtime Database)
import { initializeApp, getApps, getApp } from 'firebase/app';
import { getAnalytics } from 'firebase/analytics';
import {
  getDatabase,
  ref as dbRef,
  update as dbUpdate,
  set as dbSet,
  child as dbChild,
  get as dbGet,
  onValue as dbOnValue  // <-- add this
} from 'firebase/database';

// Firebase web config (public by design)
const firebaseConfig = {
  apiKey: 'AIzaSyDzZ0CAodwCYHgQJahsgcxJrmWiIIbVGM',
  authDomain: 'sports-live-c25ca.firebaseapp.com',
  databaseURL: 'https://sports-live-c25ca-default-rtdb.firebaseio.com',
  projectId: 'sports-live-c25ca',
  storageBucket: 'sports-live-c25ca.firebasestorage.app',
  messagingSenderId: '532399482994',
  appId: '1:532399482994:web:3d1353393886b64dc1baeb',
  measurementId: 'G-8JT6X5DF6B',
};

// Singleton app instance across HMR/dev reloads
const app = getApps().length ? getApp() : initializeApp(firebaseConfig);

// Realtime Database instance
export const db = getDatabase(app);

// Modular functions for convenience
export const ref = dbRef;
export const update = dbUpdate;
export const set = dbSet;
export const child = dbChild;
export const get = dbGet;
export const onValue = dbOnValue;  // <-- export onValue

// Guard analytics for browser-only environments
export const initAnalytics = () => {
  if (typeof window !== 'undefined' && 'measurementId' in firebaseConfig) {
    try {
      return getAnalytics(app);
    } catch (_) {
      // ignore if analytics not supported
    }
  }
  return null;
};

export default app;
