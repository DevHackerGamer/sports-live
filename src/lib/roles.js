// Client-side role utilities to interpret Clerk metadata consistently

export function getUserRoles(user) {
  if (!user) return [];
  const raw = [
    user?.privateMetadata?.type,
    user?.publicMetadata?.type,

  ]
    .concat(Array.isArray(user?.privateMetadata?.roles) ? user.privateMetadata.roles : [])
    .concat(Array.isArray(user?.publicMetadata?.roles) ? user.publicMetadata.roles : [])
    .filter(Boolean)
    .map(v => String(v).toLowerCase());
  return Array.from(new Set(raw));
}

export function isAdminFromUser(user) {
  const roles = getUserRoles(user);
  return roles.includes('admin');
}
