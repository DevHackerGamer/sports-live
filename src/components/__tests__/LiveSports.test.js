import React from 'react';
import { render, screen, fireEvent, waitFor, act } from '@testing-library/react';
import LiveSports from '../sports/LiveSports';
import * as ClerkReact from '@clerk/clerk-react';
import * as api from '../../lib/api';

// Mock Clerk hooks
jest.mock('@clerk/clerk-react', () => ({
  useUser: jest.fn(),
}));

// Mock API client
jest.mock('../../lib/api', () => ({
  apiClient: {
    getUserWatchlist: jest.fn(),
    getMatchesByDate: jest.fn(),
    getTeams: jest.fn(),
    addUserMatch: jest.fn(),
  },
}));

// Mock custom hooks
jest.mock('../../hooks/useLiveSports', () => ({
  useLiveSports: jest.fn(),
}));

// Mock child components and utilities
jest.mock('../TeamInfo/TeamInfo', () => ({ team, onBack }) => (
  <div data-testid="team-info">
    Team Info: {team?.name}
    <button onClick={onBack}>Back to Matches</button>
  </div>
));

jest.mock('../../lib/leagueNames', () => ({
  getLeagueName: jest.fn((code) => code || 'Unknown League'),
}));

// Mock FontAwesome
jest.mock('@fortawesome/react-fontawesome', () => ({
  FontAwesomeIcon: ({ icon, ...props }) => <span data-testid="font-awesome-icon" {...props} />,
}));

describe('LiveSports Component', () => {
  const mockOnMatchSelect = jest.fn();
  const mockUser = { id: 'user123', firstName: 'John' };

  const mockSportsData = {
    games: [
      {
        id: 'match1',
        homeTeam: { name: 'Arsenal' },
        awayTeam: { name: 'Chelsea' },
        homeScore: 2,
        awayScore: 1,
        status: 'live',
        competition: 'Premier League',
        competitionCode: 'PL',
        utcDate: '2024-01-15T15:00:00Z',
        venue: 'Emirates Stadium',
        matchday: 1,
        __utcMs: Date.parse('2024-01-15T15:00:00Z'),
      },
      {
        id: 'match2',
        homeTeam: { name: 'Liverpool' },
        awayTeam: { name: 'Manchester City' },
        homeScore: 0,
        awayScore: 0,
        status: 'scheduled',
        competition: 'Premier League',
        competitionCode: 'PL',
        utcDate: '2024-01-16T17:30:00Z',
        venue: 'Anfield',
        matchday: 1,
        __utcMs: Date.parse('2024-01-16T17:30:00Z'),
      },
    ],
    range: {
      dateFrom: '2024-01-15',
      dateTo: '2024-01-21',
    },
    totalMatches: 2,
    source: 'ESPN',
    environment: 'production',
  };

  const mockTeamsData = {
    data: [
      { name: 'Arsenal', crest: '/arsenal.png' },
      { name: 'Chelsea', crest: '/chelsea.png' },
      { name: 'Liverpool', crest: '/liverpool.png' },
      { name: 'Manchester City', crest: '/mancity.png' },
    ],
  };

  beforeEach(() => {
    jest.clearAllMocks();
    
    // Default mocks
    ClerkReact.useUser.mockReturnValue({ user: mockUser });
    
    // Mock useLiveSports hook
    const { useLiveSports } = require('../../hooks/useLiveSports');
    useLiveSports.mockReturnValue({
      sportsData: mockSportsData,
      isConnected: true,
      error: null,
      lastUpdated: new Date('2024-01-15T14:30:00Z'),
    });

    // Mock API responses
    api.apiClient.getUserWatchlist.mockResolvedValue({ data: [] });
    api.apiClient.getMatchesByDate.mockResolvedValue({ data: [] });
    api.apiClient.getTeams.mockResolvedValue(mockTeamsData);
    api.apiClient.addUserMatch.mockResolvedValue({});

    // Mock league names
    const { getLeagueName } = require('../../lib/leagueNames');
    getLeagueName.mockImplementation((code) => {
      const leagues = { PL: 'Premier League', UCL: 'Champions League' };
      return leagues[code] || code || 'Unknown League';
    });
  });

  it('renders loading state initially', () => {
    const { useLiveSports } = require('../../hooks/useLiveSports');
    useLiveSports.mockReturnValue({
      sportsData: null,
      isConnected: false,
      error: null,
      lastUpdated: null,
    });

    render(<LiveSports onMatchSelect={mockOnMatchSelect} />);

    expect(screen.getByTestId('loading')).toBeInTheDocument();
    expect(screen.getByText('Loading live matches...')).toBeInTheDocument();
  });

  it('renders error state when there is an error', () => {
    const { useLiveSports } = require('../../hooks/useLiveSports');
    useLiveSports.mockReturnValue({
      sportsData: null,
      isConnected: false,
      error: new Error('Connection failed'),
      lastUpdated: null,
    });

    render(<LiveSports onMatchSelect={mockOnMatchSelect} />);

    expect(screen.getByTestId('error')).toBeInTheDocument();
    expect(screen.getByText('Connection Error')).toBeInTheDocument();
    expect(screen.getByText('Unable to fetch live sports data')).toBeInTheDocument();
  });

  it('renders main content with matches when data is available', async () => {
    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    expect(screen.getByText('Live Football')).toBeInTheDocument();
    expect(screen.getByText('Live')).toBeInTheDocument();
    expect(screen.getByTestId('matches-container')).toBeInTheDocument();
  });

  it('shows connection status and current time', async () => {
    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    expect(screen.getByText('Live')).toBeInTheDocument();
    expect(screen.getByText(/Updated/)).toBeInTheDocument();
  });

  it('handles match selection when match card is clicked', async () => {
    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    const matchCards = screen.getAllByTestId(/match-/);
    fireEvent.click(matchCards[0]);

    expect(mockOnMatchSelect).toHaveBeenCalledWith(mockSportsData.games[0]);
  });

  it('switches between upcoming and past view modes', async () => {
    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    // Initially in upcoming view
    expect(screen.getByText('Upcoming')).toHaveClass('ls-active');

    // Switch to past view
    const pastButton = screen.getByText('Past Results');
    fireEvent.click(pastButton);

    expect(screen.getByText('Past Results')).toHaveClass('ls-active');
  });

  it('loads past matches when switching to past view', async () => {
    const pastMatches = [
      {
        id: 'past1',
        homeTeam: { name: 'Arsenal' },
        awayTeam: { name: 'Tottenham' },
        homeScore: 3,
        awayScore: 1,
        status: 'finished',
        competition: 'Premier League',
        competitionCode: 'PL',
        utcDate: '2024-01-14T15:00:00Z',
        venue: 'Emirates Stadium',
        matchday: 1,
        __utcMs: Date.parse('2024-01-14T15:00:00Z'),
      },
    ];

    api.apiClient.getMatchesByDate.mockResolvedValue({ data: pastMatches });

    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    // Switch to past view
    const pastButton = screen.getByText('Past Results');
    await act(async () => {
      fireEvent.click(pastButton);
    });

    await waitFor(() => {
      expect(api.apiClient.getMatchesByDate).toHaveBeenCalled();
    });
  });

  it('handles team selection when team name or crest is clicked', async () => {
    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    const teamNames = screen.getAllByText('Arsenal');
    fireEvent.click(teamNames[0]);

    // Should show TeamInfo component
    expect(screen.getByTestId('team-info')).toBeInTheDocument();
    expect(screen.getByText('Team Info: Arsenal')).toBeInTheDocument();
  });

  it('adds match to watchlist when watchlist button is clicked', async () => {
    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    // Hover over a match card to show watchlist button
    const matchCard = screen.getAllByTestId(/match-/)[0];
    fireEvent.mouseEnter(matchCard);

    const watchlistButton = await screen.findByText('+ Watchlist');
    await act(async () => {
      fireEvent.click(watchlistButton);
    });

    expect(api.apiClient.addUserMatch).toHaveBeenCalledWith(
      mockUser.id,
      mockSportsData.games[0]
    );
  });

  it('fetches user watchlist on component mount', async () => {
    const watchlistData = { data: [{ matchId: 'match1' }] };
    api.apiClient.getUserWatchlist.mockResolvedValue(watchlistData);

    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    expect(api.apiClient.getUserWatchlist).toHaveBeenCalledWith(mockUser.id);
  });

  it('handles missing user gracefully', async () => {
    ClerkReact.useUser.mockReturnValue({ user: null });

    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    // Should still render without errors
    expect(screen.getByText('Live Football')).toBeInTheDocument();
  });

  it('displays different status badges correctly', async () => {
    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    // Check for live status badge
    expect(screen.getByTestId('status-live')).toBeInTheDocument();
    expect(screen.getByTestId('status-scheduled')).toBeInTheDocument();
  });

  it('shows empty state when no matches are available', async () => {
    const { useLiveSports } = require('../../hooks/useLiveSports');
    useLiveSports.mockReturnValue({
      sportsData: { games: [], totalMatches: 0 },
      isConnected: true,
      error: null,
      lastUpdated: new Date(),
    });

    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    expect(screen.getByText('No upcoming matches in this date range')).toBeInTheDocument();
  });

  it('handles API errors gracefully', async () => {
    api.apiClient.getTeams.mockRejectedValue(new Error('Failed to fetch teams'));

    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    // Should still render matches even if team crests fail
    expect(screen.getByText('Arsenal')).toBeInTheDocument();
  });

  it('formats dates and times correctly', async () => {
    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    // Check if date range is displayed
    expect(screen.getByText(/Range:/)).toBeInTheDocument();
    expect(screen.getByText(/Matches: 2/)).toBeInTheDocument();
  });

  it('displays match scores and minute for live matches', async () => {
    await act(async () => {
      render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    });

    expect(screen.getByText('2')).toBeInTheDocument(); // Arsenal score
    expect(screen.getByText('1')).toBeInTheDocument(); // Chelsea score
  });

  it('matches snapshot with loaded data', async () => {
    await act(async () => {
      const { container } = render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
      expect(container).toMatchSnapshot();
    });
  });

  it('matches snapshot in loading state', () => {
    const { useLiveSports } = require('../../hooks/useLiveSports');
    useLiveSports.mockReturnValue({
      sportsData: null,
      isConnected: false,
      error: null,
      lastUpdated: null,
    });

    const { container } = render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    expect(container).toMatchSnapshot();
  });

  it('matches snapshot in error state', () => {
    const { useLiveSports } = require('../../hooks/useLiveSports');
    useLiveSports.mockReturnValue({
      sportsData: null,
      isConnected: false,
      error: new Error('Test error'),
      lastUpdated: null,
    });

    const { container } = render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
    expect(container).toMatchSnapshot();
  });

  describe('TeamInfo integration', () => {
    it('shows TeamInfo when a team is selected', async () => {
      await act(async () => {
        render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
      });

      const teamNames = screen.getAllByText('Arsenal');
      fireEvent.click(teamNames[0]);

      expect(screen.getByTestId('team-info')).toBeInTheDocument();
      expect(screen.getByText('Back to Matches')).toBeInTheDocument();
    });

    it('returns to matches view when back button is clicked in TeamInfo', async () => {
      await act(async () => {
        render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
      });

      // Select a team
      const teamNames = screen.getAllByText('Arsenal');
      fireEvent.click(teamNames[0]);

      // Click back button
      const backButton = screen.getByText('Back to Matches');
      fireEvent.click(backButton);

      expect(screen.queryByTestId('team-info')).not.toBeInTheDocument();
    });
  });

  describe('Watchlist functionality', () => {
    it('shows watchlist button only when user is logged in', async () => {
      ClerkReact.useUser.mockReturnValue({ user: null });

      await act(async () => {
        render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
      });

      const matchCard = screen.getAllByTestId(/match-/)[0];
      fireEvent.mouseEnter(matchCard);

      expect(screen.queryByText('+ Watchlist')).not.toBeInTheDocument();
    });

    it('does not show watchlist button for already added matches', async () => {
      const watchlistData = { data: [{ matchId: 'match1' }] };
      api.apiClient.getUserWatchlist.mockResolvedValue(watchlistData);

      await act(async () => {
        render(<LiveSports onMatchSelect={mockOnMatchSelect} />);
      });

      const matchCard = screen.getAllByTestId(/match-/)[0];
      fireEvent.mouseEnter(matchCard);

      expect(screen.queryByText('+ Watchlist')).not.toBeInTheDocument();
    });
  });
});