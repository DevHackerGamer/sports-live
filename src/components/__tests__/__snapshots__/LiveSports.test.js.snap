// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiveSports Component matches snapshot in error state 1`] = `
<div>
  <div
    class="ls-live-sports ls-error-state"
    data-testid="error"
  >
    <div
      class="ls-error-content"
    >
      <h3>
        Connection Error
      </h3>
      <p>
        Unable to fetch live sports data
      </p>
      <div
        style="display: none;"
      >
        Failed to load matches
      </div>
    </div>
  </div>
</div>
`;

exports[`LiveSports Component matches snapshot in loading state 1`] = `
<div>
  <div
    class="ls-live-sports ls-loading-state"
    data-testid="loading"
  >
    <div
      class="ls-loading-content"
    >
      <div
        class="ls-loading-spinner"
      />
      <h3>
        Loading live matches...
      </h3>
    </div>
  </div>
</div>
`;

exports[`LiveSports Component matches snapshot with loaded data 1`] = `<div />`;
